csharp/test/Ice/interceptor/AllTests.cs

```diff
@@ -0,0 +1,293 @@
+
+            output.Write("testing retry... ");
+            output.Flush();
+            TestHelper.Assert(prx.AddWithRetry(33, 12) == 45);
+            output.WriteLine("ok");
+
+            output.Write("testing remote exception... ");
+            output.Flush();
+            try
+            {
+                prx.BadAdd(33, 12);
+                TestHelper.Assert(false);
+            }
+            catch (InvalidInputException)
+            {
+                // expected
+            }
+            output.WriteLine("ok");
+
+            output.Write("testing ONE... ");
+            output.Flush();
+            try
+            {
+                prx.NotExistAdd(33, 12);
+                TestHelper.Assert(false);
+            }
+            catch (ObjectNotExistException)
+            {
+                // expected
+            }
+            output.WriteLine("ok");
+
+            output.Write("testing exceptions raised by the interceptor... ");
+            output.Flush();
+            var exceptions = new List<(string operation, string kind)>
+            {
+                ("raiseBeforeDispatch", "invalidInput"),
+                ("raiseBeforeDispatch", "notExist"),
+                ("raiseAfterDispatch", "invalidInput"),
+                ("raiseAfterDispatch", "notExist")
+            };
+
+            foreach ((string operation, string kind) in exceptions)
+            {
+                var ctx = new Dictionary<string, string>
+                {
+                    { operation, kind }
+                };
+
+                try
+                {
+                    prx.IcePing(ctx);
+                    TestHelper.Assert(false);
+                }
+                catch (InvalidInputException) when (kind == "invalidInput")
+                {
+                }
+                catch (ObjectNotExistException) when (kind == "notExist")
+                {
+                }
+            }
+            output.WriteLine("ok");
+
+            output.Write("testing invocation interceptors... ");
+            output.Flush();
+            {
+                using var communicator = new Communicator(prx.Communicator.GetProperties());
+                communicator.InterceptInvocation((target, request, next) =>
+                    {
+                        request.Context["interceptor-1"] = "interceptor-1";
+                        if (ice2)
+                        {
+                            request.AddBinaryContextEntry(110, 110, (ostr, v) => ostr.WriteInt(v));
+                        }
+                        return next(target, request);
+                    });
+
+                communicator.InterceptInvocation(async (target, request, next) =>
+                    {
+                        TestHelper.Assert(request.Context["interceptor-1"] == "interceptor-1");
+                        request.Context["interceptor-2"] = "interceptor-2";
+                        if (ice2)
```

Comment:
I understand in the tests we know if it's ice1 or ice2, but that's not true in general or interceptors, you might have a mix. It would be nice to use the interceptor parameters to figure out the protocol.

Reply:
right this should be `request.Protocol == Protocol.Ice2`